crates/bevy_reflect/src/path/parse.rs

```diff
@@ -114,33 +127,46 @@ impl<'a> Ident<'a> {
     }
 }
 
+// NOTE: We use repr(u8) so that the `match byte` in `Token::symbol_from_byte`
+// becomes a "check `byte` is one of SYMBOLS and forward its value" this makes
+// the optimizer happy, and shaves off a few cycles.
 #[derive(Debug, PartialEq, Eq)]
+#[repr(u8)]
 enum Token<'a> {
-    Dot,
-    Pound,
-    OpenBracket,
-    CloseBracket,
+    Dot = b'.',
+    Pound = b'#',
+    OpenBracket = b'[',
+    CloseBracket = b']',
     Ident(Ident<'a>),
 }
 impl fmt::Display for Token<'_> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            Token::Dot => f.write_str("."),
-            Token::Pound => f.write_str("#"),
-            Token::OpenBracket => f.write_str("["),
-            Token::CloseBracket => f.write_str("]"),
-            Token::Ident(ident) => f.write_str(ident.0),
+        if let Token::Ident(ident) = self {
+            f.write_str(ident.0)
+        } else {
+            let byte = match self {
+                Token::Dot => b'.',
+                Token::Pound => b'#',
+                Token::OpenBracket => b'[',
+                Token::CloseBracket => b']',
+                Token::Ident(_) => unreachable!(),
+            };
+            let byte = &[byte];
+            // SAFETY: we just defined `byte`, it is indeed a single ASCII character,
+            // therefore valid utf8.
+            let str = unsafe { from_utf8_unchecked(byte) };
+            f.write_str(str)
```

Comment:
i think this change is much worse for clarity and unsafe footprint. we should revert it assuming it doesn't have performance concerns.

Reply:
yeah, just for a fmt::Display impl, it's probably worth getting rid of.